wp-includes/class-wp-service-workers.php

```diff
@@ -63,6 +135,119 @@ public function init() {
 		do_action_ref_array( 'wp_default_service_workers', array( &$this ) );
 	}
 
+	/**
+	 * Get workbox script.
+	 *
+	 * @return string Script.
+	 */
+	public function get_workbox_script() {
+
+		$workbox_dir = 'wp-includes/js/workbox-v3.4.1/';
+
+		$script = sprintf(
+			"importScripts( %s );\n",
+			wp_json_encode( PWA_PLUGIN_URL . $workbox_dir . 'workbox-sw.js', 64 /* JSON_UNESCAPED_SLASHES */ )
+		);
+
+		$options = array(
+			'debug'            => WP_DEBUG,
+			'modulePathPrefix' => PWA_PLUGIN_URL . $workbox_dir,
+		);
+		$script .= sprintf( "workbox.setConfig( %s );\n", wp_json_encode( $options, 64 /* JSON_UNESCAPED_SLASHES */ ) );
+
+		/**
+		 * Filters whether navigation preload is enabled.
+		 *
+		 * The filtered value will be sent as the Service-Worker-Navigation-Preload header value if a truthy string.
+		 * This filter should be set to return false to disable navigation preload such as when a site is using
+		 * the app shell model.
+		 *
+		 * @param bool|string $navigation_preload Whether to use navigation preload.
+		 */
+		$navigation_preload = apply_filters( 'service_worker_navigation_preload', true ); // @todo This needs to vary between admin and backend.
+		if ( false !== $navigation_preload ) {
+			if ( is_string( $navigation_preload ) ) {
+				$script .= sprintf( "workbox.navigationPreload.enable( %s );\n", wp_json_encode( $navigation_preload ) );
+			} else {
+				$script .= "workbox.navigationPreload.enable();\n";
+			}
+		} else {
+			$script .= "/* Navigation preload disabled. */\n";
+		}
+
+		return $script;
+	}
+
+	/**
+	 * Add precaching for wp-admin and wp-includes .js and .css files.
+	 */
+	protected function precache_admin_assets() {
+
+		$admin_dir    = ABSPATH . 'wp-admin/';
+		$admin_images = list_files( $admin_dir . 'images/' );
+		$inc_images   = list_files( ABSPATH . WPINC . '/images/' );
+
+		$routes = array_merge(
+			$this->get_routes_from_file_list( $admin_images, 'wp-admin' ),
+			$this->get_routes_from_file_list( $inc_images, 'wp-includes' ),
+			$this->get_admin_routes_from_dependency_list( wp_scripts()->registered ),
+			$this->get_admin_routes_from_dependency_list( wp_styles()->registered )
+		);
+
+		if ( empty( $routes ) ) {
+			return;
+		}
+
+		$this->register_precached_routes( $routes );
+	}
+
+	/**
+	 * Get routes from dependency list.
+	 *
+	 * @param array $dependencies Array of _WP_Dependency objects.
+	 * @return array Array of routes.
+	 */
+	protected function get_admin_routes_from_dependency_list( $dependencies ) {
+		$routes = array();
+		foreach ( $dependencies as $handle => $params ) {
+
+			// Only precache scripts from wp-admin and wp-includes.
+			if ( false === strpos( $params->src, 'wp-admin' ) && false === strpos( $params->src, 'wp-includes' ) ) {
+				continue;
+			}
+			$revision = false === $params->ver ? get_bloginfo( 'version' ) : $params->ver;
```

Comment:
We could potentially opt-in to using the service worker for theme/plugin assets if we detect that the script is located within `wp-content/themes` or `wp-content/plugins` and then grab the version for the theme or plugin to use if the `ver` is empty. Or even, the `ver` could be combined with the version of the plugin to use as the revision.

Reply:
Maybe precaching theme/plugin assets could be something that could be opted in from admin UI as a setting instead of doing this by default? Thinking that precaching theme and plugin files could cause issues (as you mentioned above) if the version is not changed when updating the code (either in WP repo or directly in the environment). Feels like it's better to leave the asset precaching by default for the admin interface only, at least for now. Thoughts?